File: src/main/kotlin/kotlin99/P13.kt
package kotlin99

/**
 * Created by rick on 2017/2/18.
 */

fun <T> encodeDirect(list: List<T>): List<Pair<Int, T>> {
    if (list.isEmpty())
        return emptyList()
    else {
        if (list.size == 1)
            return listOf(Pair(1, list[0]))
        else {
            var result = list.takeWhile { it == list[0] }
            return listOf(Pair(result.size, result[0])) + encodeDirect(list.drop(result.size))
        }
    }
}

File: src/main/scala/scala99/P13.scala
package scala99

import org.scalatest.{FlatSpec, Matchers}

/**
  * Created by rick on 3/31/2017.
  */
object P13 {
    def encodeDirect[T](list: List[T]): List[(Int, T)] = {
        if (list.isEmpty)
            Nil
        else {
            val (before, after) = list.span(_ == list.head)
            (before.length, before.head) +: encodeDirect(after)
        }
    }

}

class P13Spec extends FlatSpec with Matchers {
    def fixture = new {
        val list = List('a, 'a, 'a, 'a, 'b, 'c, 'c, 'a, 'a, 'd, 'e, 'e, 'e, 'e)
        val result = List((4, 'a), (1, 'b), (2, 'c), (2, 'a), (1, 'd), (4, 'e))
    }

    it should "return run-length encoding" in {
        P13.encodeDirect(fixture.list) should contain theSameElementsInOrderAs (fixture.result)
    }
}
